File: cpp/CyberSheep.cpp
#include "CyberSheep.h"
#include <cmath>
#include <iostream>

CyberSheep::CyberSheep(int X, int Y, World* world) :Animal(CYBERSHEEP, 11, 4, X, Y, world) {};

void CyberSheep::Action()
{
	Position hogweedP = Position(-1, -1);
	int shortest = GetWorld()->GetSizeX() + GetWorld()->GetSizeY();
	for (int i = 0; i < GetWorld()->GetSizeX();i++)
		for (int j = 0; j < GetWorld()->GetSizeY(); j++)
		{
			if ((Grid(i,j))->GetName() == HOGWEED)
			{
				int d = abs(i - this->GetPosition().X) + abs(j - this->GetPosition().Y);
				if (d < shortest)
				{
					shortest = d;
					hogweedP = Position(i, j);

					std::cout << "'" << GetName() << "'(" << GetPosition().X << "," << GetPosition().Y << ") find '" << (Grid(hogweedP.X,hogweedP.Y))->GetName() << "'(" << hogweedP.X << "," << hogweedP.Y << ")" << std::endl;
				}
			}
		}

	Position defenderP(-1, -1);
	Organism* attacker = this;

	//if there is hogweed
	if (hogweedP.X != -1)
	{
		if (hogweedP.X > this->GetPosition().X)
			defenderP = GetPosition().Right(GetWorld());
		else if (hogweedP.X < this->GetPosition().X)
			defenderP = GetPosition().Left(GetWorld());
		else if (hogweedP.Y < this->GetPosition().Y)
			defenderP = GetPosition().Up(GetWorld());
		else if (hogweedP.Y > this->GetPosition().Y)
			defenderP = GetPosition().Down(GetWorld());
	}
	else	//no hog weed
	{
		int direction = rand() % 4;

		switch (direction)
		{
		case 0:	//left
			defenderP = GetPosition().Left(GetWorld());
			//this->SetPosition(GetPosition().Left(GetWorld()));
			break;
		case 1:	//right
			defenderP = GetPosition().Right(GetWorld());
			break;
		case 2:	//up
			defenderP = GetPosition().Up(GetWorld());
			break;
		case 3:	//down
			defenderP = GetPosition().Down(GetWorld());
			break;
		}
	}

	//if they are same animal
	if ((Grid(defenderP.X, defenderP.Y))->GetName() == this->GetName())
	{
		//baby is born next to them
		Position babyPosition(-1, -1);
		defenderP.Y >= this->GetPosition().Y ?
			babyPosition = Position(defenderP.X, defenderP.Y + 1) :
			babyPosition = Position(this->GetPosition().X, this->GetPosition().Y + 1);

		if (babyPosition.Y >= GetWorld()->GetSizeY())
			babyPosition.Y = 0;

		std::cout << "'" << GetName() << "'(" << GetPosition().X << "," << GetPosition().Y << ") and '" << (Grid(defenderP.X, defenderP.Y))->GetName() << "'(" << defenderP.X << "," << defenderP.Y << ") has a baby." << std::endl;

		//baby's coliision
		/*if ((Grid(babyPosition.X, babyPosition.Y))->GetName() == this->GetName())
		return;*/
		//else
		//{
			attacker = new Organism(*this);
			attacker->SetPosition(Position(-1, -1));
			(Grid(defenderP.X, defenderP.Y))->Collision(attacker);
			//delete(attacker);
	//}
	}
	else
	//if they are different animal
	{
		std::cout << "'" << GetName() << "'(" << GetPosition().X << "," << GetPosition().Y << ") attack '" << (Grid(defenderP.X, defenderP.Y))->GetName() << "'(" << defenderP.X << "," << defenderP.Y << ")" << std::endl;

		Grid(GetPosition().X, GetPosition().Y) = new nullGrid(GetPosition());
		(Grid(defenderP.X, defenderP.Y))->Collision(attacker);
	}
}

File: java/src/eti/Project2/Organisms/CyberSheep.java
package eti.Project2.Organisms;

import eti.Project2.GUI;
import eti.Project2.MapPosition;
import eti.Project2.World;

import java.util.Random;

/**
 * Created By IDEA
 * Author: s188026 Yifei Liu
 * Date: 2021/5/30 13:06
 * Description: CyberSheep  'c'
 * Its main goal is the extermination of Sosnowsky's hogweed.
 * It always moves towards the closes hogweed and tries to eat it.
 * If there are no Sosnowsky's hogweeds, it behaves like a normal sheep.
 * Eats Sosnowsky's hogweed.
 */

public class CyberSheep extends Animal {
    //constructor
    public CyberSheep(char n, int s, int i, int X, int Y, World w) {
        super(n, s, i, X, Y, w);
    }

    public CyberSheep(CyberSheep c) {
        super(c);
    }

    public CyberSheep(int X, int Y, World world) {
        setName('c');
        setStrength(11);
        setInitiative(4);
        setPosition(new MapPosition(X, Y));
        setWorld(world);
    }

    @Override
    public void Action() {
        MapPosition hogweedP = new MapPosition(-1, -1);
        int shortest = getWorld().getSizeX() + getWorld().getSizeY();
        for (int i = 0; i < getWorld().getSizeX(); i++) {
            for (int j = 0; j < getWorld().getSizeY(); j++) {
                if (getWorld().organisms[j][i].getName() == '%') {
                    int d = Math.abs(i - this.getPosition().getX()) + Math.abs(j - this.getPosition().getY());
                    if (d < shortest) {
                        shortest = d;
                        hogweedP = new MapPosition(i, j);

                        /**std::cout + "'" + getName() + "'(" + getPosition().getX() + "," + getPosition().getY() + ") find '" + (Grid(hogweedP.getX(),hogweedP.getY())).getName() + "'(" + hogweedP.getX() + "," + hogweedP.getY() + ")" + std::endl;*/
                        GUI.getInformation().append("'" + getName() + "'(" + getPosition().getX() + "," + getPosition().getY() + ") find '" + getWorld().organisms[hogweedP.getY()][hogweedP.getX()].getName() + "'(" + hogweedP.getX() + "," + hogweedP.getY() + ")\n");
                    }
                }
            }
        }

        MapPosition defenderP = new MapPosition(-1, -1);
        Organism attacker = this;

        //if there is hogweed
        if (hogweedP.getX() != -1) {
            if (hogweedP.getX() > this.getPosition().getX())
                defenderP = getPosition().Right(getWorld());
            else if (hogweedP.getX() < this.getPosition().getX())
                defenderP = getPosition().Left(getWorld());
            if (hogweedP.getY() < this.getPosition().getY())
                defenderP = getPosition().Up(getWorld());
            else if (hogweedP.getY() > this.getPosition().getY())
                defenderP = getPosition().Down(getWorld());
        }

        //no hogweed
        else
        {
            Random random = new Random();
            int direction = random.nextInt(8);

            switch (direction) {
                case 0:    //left
                    defenderP = getPosition().Left(getWorld());
                    //this.SetPosition(getPosition().Left(getWorld()));
                    break;
                case 1:    //right
                    defenderP = getPosition().Right(getWorld());
                    break;
                case 2:    //up
                    defenderP = getPosition().Up(getWorld());
                    break;
                case 3:    //down
                    defenderP = getPosition().Down(getWorld());
                    break;
                case 4:    //left down
                    defenderP = getPosition().LeftDown(getWorld());
                    break;
                case 5:    //left up
                    defenderP = getPosition().LeftUp(getWorld());
                    break;
                case 6:    //right down
                    defenderP = getPosition().RightDown(getWorld());
                    break;
                case 7:    //right up
                    defenderP = getPosition().RightUp(getWorld());
                    break;
            }
        }

        //if they are same animal
        if (getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == this.getName()) {
            //baby is born next to them
            MapPosition babyPosition =
                    defenderP.getY() >= this.getPosition().getY() ?
                            new MapPosition(defenderP.getX(), defenderP.getY() + 1) :
                            new MapPosition(this.getPosition().getX(), this.getPosition().getY() + 1);

            if (babyPosition.getY() >= getWorld().getSizeY())
                babyPosition.setY(0);

            /**std::cout + "'" + getName() + "'(" + getPosition().getX() + "," + getPosition().getY() + ") and '" + (Grid(defenderP.getX(), defenderP.getY())).getName() + "'(" + defenderP.getX() + "," + defenderP.getY() + ") has a baby." + std::endl;*/
            GUI.getInformation().append("'" + getName() + "'(" + getPosition().getX() + "," + getPosition().getY() + ") and '" + getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() + "'(" + defenderP.getX() + "," + defenderP.getY() + ") has a cyber baby.\n");


            //baby's collision
            attacker = new CyberSheep(this);
            attacker.setPosition(new MapPosition(-1, -1));
        } else
        //if they are different animal
        {
            /**std::cout + "'" + getName() + "'(" + getPosition().getX() + "," + getPosition().getY() + ") attack '" + (Grid(defenderP.getX(), defenderP.getY())).getName() + "'(" + defenderP.getX() + "," + defenderP.getY() + ")" + std::endl;*/
            char defenderName = getWorld().organisms[defenderP.getY()][defenderP.getX()].getName();
            if(defenderName != '_')
                GUI.getInformation().append("'" + getName() + "'(" + getPosition().getX() + "," + getPosition().getY() + ") attack '" + defenderName + "'(" + defenderP.getX() + "," + defenderP.getY() + ")\n");
            else
                GUI.getInformation().append("'" + getName() + "'(" + getPosition().getX() + "," + getPosition().getY() + ") move to (" + defenderP.getX() + "," + defenderP.getY() + ")\n");

            getWorld().organisms[getPosition().getY()][getPosition().getX()] = new Organism('_', -1, -1, getPosition().getX(), getPosition().getY(), getWorld());
        }
        getWorld().organisms[defenderP.getY()][defenderP.getX()].Collision(attacker);
    }
}
